File: java/src/org/openstreetmap/util/Node.java
/*
 * Copyright (C) 2005 Tom Carden (tom@somethingmodern.com)
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place - Suite 330, Boston, MA 02111-1307, USA.
 *  
 */

package org.openstreetmap.util;

import java.util.Collection;
import java.util.Map;
import java.util.Vector;

import org.openstreetmap.client.Tile;
import org.openstreetmap.processing.OsmApplet;

/**
 * Minimal representation of OpenStreetMap node (lat/lon pair, with uid)
 */
public class Node extends OsmPrimitive {

	/**
	 * The coordinates of this node
	 */
	public Point coor;
	
	/**
	 * All lines in this node.
	 * Type: Line
	 */
	public Collection lines = new Vector();

	/**
	 * Create the node from projected values.
	 */
	public Node(float x, float y, Tile projection) {
		coor = new Point(x, y, projection);
	}

	/**
	 * Create the node with all information given.
	 */
	public Node(double lat, double lon, long id) {
		coor = new Point(lat, lon);
		this.id = id;
	}

	public String toString() {
		return "[Node " + id + " lat:" + coor.lat + " lon:" + coor.lon + "]";

	}

	public String key() {
		return key(id);
	}

    public static String key(long id) {
	    return "node_" + id;
	}

	/**
	 * Note, that it is almost anytime the SQUARED distance is what you need.
	 * It is faster to calculate, does not suffer as much from rounding issues and
	 * is equal good when it comes to comparing to other squared distances.
	 * 
	 * @return The squared distance to x,y.
	 */
	public float distanceSq(float x, float y) {
		final float xabs = Math.abs(coor.x-x);
		final float yabs = Math.abs(coor.y-y);
		return xabs*xabs + yabs*yabs;
	}

	public Map getMainTable(OsmApplet applet) {
		return applet.nodes;
	}

	public String getTypeName() {
		return "node";
	}

	public String getName() {
		String name = (String)tags.get("name");
		return name!=null ? name : "";
	}

	public void register() {}
	public void unregister() {}

	protected void doCopyFrom(OsmPrimitive other) {
		coor = ((Node)other).coor;
	}
}

File: editors/osm-editor/Node.cpp
#include "Node.h"
#include "NodeMetaDataHandler.h"

/*
    Copyright (C) 2005 Nick Whitelegg, Hogweed Software, nick@hogweed.org

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA

 */
namespace OpenStreetMap
{

// 180306 updated for 0.3
int Node::toOSM(QTextStream& outfile, bool allUid)
{
    int sent_id = (osm_id>0 || (allUid&&osm_id)) ? osm_id: 0;
	outfile.precision(10); // 250306 in response to request
    outfile << "<node lat='" <<  lat <<
                "' lon='" <<  lon << "'";
    
    outfile << " id='" << sent_id << "' >" << endl;
    
    if(name != "")
        outfile << "<tag k='name' v='" << name << "'/>" << endl;

	// 180506 handle new style keys (ie. not class)
    if(type != "")
	{
		NodeMetaDataHandler mdh;
		NodeMetaData md = mdh.getMetaData(type);
		if(md.key!="" || md.value!="")
        	outfile << "<tag k='"<<md.key<<"' v='" << md.value << "'/>" << endl;
	}
    outfile << "</node>" << endl;
}

// 180306 updated for 0.3
QByteArray Node::toOSM()
{
    QByteArray xml;
    QTextStream str(xml, IO_WriteOnly);
    str<<"<osm version='0.3'>"<<endl;
    toOSM(str);
    str<<"</osm>"<<endl;
    str<<'\0';
    return xml;
}
// 180306 not changed as this is the old curl way of doing it
// 130506 took out the old curl way of doing it

// Used when creating segments from trackpoints
void Node::trackpointToNode()
{
    if(type=="trackpoint")
        type="node";
}

}
